### PR TITLE
Added support for pagination of WFS requests

### PR DESCRIPTION
**Pull request by florianesser**
refs #188

### CODE REVIEW

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+		/**
+		 * @see java.util.Iterator#hasNext()
+		 */
+		@Override
+		public boolean hasNext() {
+			if (isClosed()) {
+				return false;
+			}
+
+			if (maxNumberOfFeatures != UNLIMITED && totalFeaturesProcessed >= maxNumberOfFeatures) {
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java]: I did not see the place where we stop issuing new requests when there are no more features though we still did not hit the max number of features. Can you please point me to it?

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/FeatureCollectionHelper.groovy [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution. If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Contributors:
+ *     wetransform GmbH <http://www.wetransform.to>
+ */
+
+package eu.esdihumboldt.hale.io.gml.reader.internal.wfs
+
+import org.geotools.data.wfs.protocol.wfs.WFSException
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/FeatureCollectionHelper.groovy]: Should the WFSException class that you defined be used here?

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/StreamGmlReader.java [hunk 3]
```diff
@@ -58,6 +61,13 @@
 	 */
 	public static final String PARAM_IGNORE_NAMESPACES = "ignoreNamespaces";
 
+	/**
+	 * The name of the parameter specifying the maximum number of features to
+	 * retrieve per single WFS GetFeature request. Only useful if source is a
+	 * WFS GetFeature request URI.
+	 */
+	public static final String PARAM_FEATURES_PER_WFS_REQUEST = "featuresPerWfsRequest";
```
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/StreamGmlReader.java]: The parameter needs also to be listed in the respective provider extensions that register readers based on StreamGmlReader.

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java (hunk 1 above)
Comment (florianesser) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java]: In hasNext(), if iterator.hasNext() (the hasNext() of the GmlInstanceIterator of the current request) returns false, proceedOrClose() is called. proceedOrClose() then creates a new iterator for the next request. If the next request does not return results, close() is called. After the call to close() all calls to hasNext() will return false (and calls to next() will throw an exception).

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/FeatureCollectionHelper.groovy (hunk 2 above)
Comment (florianesser) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/FeatureCollectionHelper.groovy]: Of course, thanks

File: io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java (hunk 1 above)
Comment (stempler) [inline io/plugins/eu.esdihumboldt.hale.io.gml/src/eu/esdihumboldt/hale/io/gml/reader/internal/wfs/WfsBackedGmlInstanceCollection.java]: Ah, thanks, found it.

**Pull request merged**